core/services/ocr2/plugins/ocr2keeper/evm21/mercury/streams/streams.go

```diff
@@ -0,0 +1,303 @@
+		return
+	}
+
+	// Try to decode the revert error into streams lookup format. User upkeeps can revert with any reason, see if they
+	// tried to call mercury
+	lookupLggr.Infof("at block %d upkeep %s trying to DecodeStreamsLookupRequest performData=%s", block, upkeepId, hexutil.Encode(checkResults[i].PerformData))
+	streamsLookupErr, err := s.packer.DecodeStreamsLookupRequest(checkResult.PerformData)
+	if err != nil {
+		lookupLggr.Debugf("at block %d upkeep %s DecodeStreamsLookupRequest failed: %v", block, upkeepId, err)
+		// user contract did not revert with StreamsLookup error
+		return
+	}
+	streamsLookupResponse := &mercury.StreamsLookup{StreamsLookupError: streamsLookupErr}
+
+	if len(streamsLookupResponse.Feeds) == 0 {
+		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonInvalidRevertDataInput)
+		lookupLggr.Debugf("at block %s upkeep %s has empty feeds array", block, upkeepId)
+		return
+	}
+
+	// mercury permission checking for v0.3 is done by mercury server
+	if streamsLookupResponse.IsMercuryV02() {
+		// check permission on the registry for mercury v0.2
+		opts := s.buildCallOpts(ctx, block)
+		if state, reason, retryable, allowed, err := s.allowedToUseMercury(opts, upkeepId.BigInt()); err != nil {
+			lookupLggr.Warnf("at block %s upkeep %s failed to query mercury allow list: %s", block, upkeepId, err)
+			checkResults[i].PipelineExecutionState = uint8(state)
+			checkResults[i].IneligibilityReason = uint8(reason)
+			checkResults[i].Retryable = retryable
+			return
+		} else if !allowed {
+			lookupLggr.Debugf("at block %d upkeep %s NOT allowed to query Mercury server", block, upkeepId)
+			checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonMercuryAccessNotAllowed)
+			return
+		}
+	} else if streamsLookupResponse.IsMercuryVersionUnkown() {
+		// if mercury version cannot be determined, set failure reason
+		lookupLggr.Debugf("at block %d upkeep %s NOT allowed to query Mercury server", block, upkeepId)
+		checkResults[i].IneligibilityReason = uint8(mercury.MercuryUpkeepFailureReasonInvalidRevertDataInput)
+		return
+	}
+
+	streamsLookupResponse.UpkeepId = upkeepId.BigInt()
+	// the block here is exclusively used to call checkCallback at this block, not to be confused with the block number
+	// in the revert for mercury v0.2, which is denoted by time in the struct bc starting from v0.3, only timestamp will be supported
+	streamsLookupResponse.Block = uint64(block.Int64())
+	lookupLggr.Infof("at block %d upkeep %s DecodeStreamsLookupRequest feedKey=%s timeKey=%s feeds=%v time=%s extraData=%s", block, upkeepId, streamsLookupResponse.FeedParamKey, streamsLookupResponse.TimeParamKey, streamsLookupResponse.Feeds, streamsLookupResponse.Time, hexutil.Encode(streamsLookupResponse.ExtraData))
+	lookups[i] = streamsLookupResponse
+}
+
+func (s *streams) doLookup(ctx context.Context, wg *sync.WaitGroup, lookup *mercury.StreamsLookup, i int, checkResults []ocr2keepers.CheckResult) {
```

Comment:
Does it make sense to change `doLookup` so that it returns an error? So instead of logging error messages and having empty returns in this function, drop the logs from this function, and return an error derived from the log message itself, and then in the calling function, check if the error returned is not nil, and if so, log the error message?

Reply:
I like your approach. Returning error is helpful to debugging script, so the script can capture the error.

Reply:
I am going to return an error from doLookup and also keep the log there. Will evaluate a solid approach to capture and log the error when working on the debugging script.